dynamite_nsm/services/kibana/package/manager.py

```diff
@@ -0,0 +1,784 @@
+
+        Args:
+            username (str): kibana auth usrname
+            password (str): kibana auth passwd
+            kibana_objects_file (IO[AnyStr]): the file to parse and install
+            space (Optional[str], optional): id of the space to install the object to. Defaults to None.
+            overwrite (Optional[bool], optional): overwrite existing ids?. Defaults to True.
+            create_copies (Optional[bool], optional): create copies if an object exists with the same id?. Defaults to False.
+
+        Raises:
+            ValueError: Something went wrong
+
+        Returns:
+            dict: Response from kibana
+        """
+
+        auth = self._get_kibana_auth_securely(username, password)
+
+        if space:
+            url = f'{self.kibana_url}/s/{space}/api/saved_objects/_import'
+        else:
+            url = f'{self.kibana_url}/api/saved_objects/_import'
+        if all([overwrite, create_copies]):
+            raise ValueError("createNewCopies and overwrite cannot be used together.")
+
+        params = {'overwrite': overwrite, 'createNewCopies': create_copies}
+
+        # TODO: Catch connection denied when kibana is down and handle/inform user gracefully
+        if isinstance(kibana_objects_file, IOBase):
+            reqdata = {'file': ('dynamite_import.ndjson', kibana_objects_file)}
+        else:
+            reqdata = {'file': kibana_objects_file}
+        resp = requests.post(url, params=params, auth=auth, files=reqdata, headers={'kbn-xsrf': 'true'})
+        if resp.status_code not in range(200, 299):
+            self.logger.error(f'Kibana endpoint returned a {resp.status_code} - {resp.text}')
+            # TODO raise exception
+        return resp.json()
+
+    def install(self, package_install_path: str, username: Optional[str] = None, password: Optional[str] = None):
+        """Install a package. A package can be given as an archive or directory. A package must contain one or more ndjson files and a manifest.json
+
+        Args:
+            package_install_path (str): path to the file or folder of files
+            username (Optional[str], optional): kibana auth usrname. Defaults to None.
+            password (Optional[str], optional): kibana auth password. Defaults to None.
+        """
+        is_folder = os.path.isdir(package_install_path)
+        filepaths = []
+        if not is_folder:
+            filepaths = [package_install_path]
+        else:
+            # check for installable items by extension, they will be verified by mimetype later.
+            acceptable_extensions = ['tar.xz', 'tar.gz']
+            for itm in os.listdir(package_install_path):
+                for ex in acceptable_extensions:
+                    if itm.endswith(ex):
+                        filepaths.append(f"{package_install_path}{itm}")
+                        break
+            print(f"Found {len(filepaths)} packages to install.")
+        for _filepath in filepaths:
```

Comment:
Might make sense to break out the conditionals in this for loop into separate sub-functions. It's a bit hard to follow.